src/YieldDCA.sol

```diff
@@ -0,0 +1,567 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.19;
+
+import "forge-std/Test.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+import {IERC20} from "openzeppelin-contracts/interfaces/IERC20.sol";
+import {IERC4626} from "openzeppelin-contracts/interfaces/IERC4626.sol";
+import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {MockERC4626} from "solmate/test/utils/mocks/MockERC4626.sol";
+import {MockERC20} from "solmate/test/utils/mocks/MockERC20.sol";
+
+contract YieldDCA {
+    using FixedPointMathLib for uint256;
+    using SafeERC20 for IERC4626;
+    using SafeERC20 for IERC20;
+
+    struct Deposit {
+        uint256 shares;
+        uint256 principal;
+        uint256 epoch;
+    }
+
+    struct EpochInfo {
+        uint256 yieldSpent;
+        uint256 dcaPrice;
+        uint256 pricePerShare;
+    }
+
+    uint256 public constant DCA_PERIOD = 2 weeks;
+
+    IERC20 public dcaToken;
+    IERC4626 public vault;
+    ISwapper public swapper;
+
+    uint256 public currentEpoch = 1; // starts from 1
+    uint256 public currentEpochTimestamp = block.timestamp;
+    uint256 public totalPrincipalDeposted;
+    mapping(address => Deposit) public deposits;
+    mapping(uint256 => EpochInfo) public epochDetails;
+
+    constructor(IERC20 _dcaToken, IERC4626 _vault, ISwapper _swapper) {
+        dcaToken = _dcaToken;
+        vault = _vault;
+        swapper = _swapper;
+
+        // approve swapper to spend deposits on DCA token
+        IERC20(vault.asset()).approve(address(swapper), type(uint256).max);
+    }
+
+    function deposit(uint256 _amount) external {
+        uint256 principal = vault.convertToAssets(_amount);
+        vault.safeTransferFrom(msg.sender, address(this), _amount);
+
+        Deposit storage position = deposits[msg.sender];
+
+        // check if user has already deposited in the past
+        if (position.epoch != 0 && position.epoch < currentEpoch) {
+            (uint256 shares, uint256 dcaTokens) = _calculateBalances(position);
+            position.shares = shares;
+
+            if (dcaTokens != 0) {
+                dcaToken.safeTransfer(msg.sender, dcaTokens);
+            }
+        }
+
+        position.shares += _amount;
+        position.principal += principal;
+        position.epoch = currentEpoch;
+
+        totalPrincipalDeposted += principal;
+    }
+
+    function executeDCA() external {
+        if (block.timestamp < currentEpochTimestamp + DCA_PERIOD) return;
+
+        uint256 yieldInShares = calculateCurrentYieldInShares();
+
+        // TODO: or revert if yield is 0?
+        if (yieldInShares == 0) return;
+
+        uint256 yield = vault.redeem(yieldInShares, address(this), address(this));
+        // TODO: use asset.balanceOf here instead of yield?
+
+        uint256 realizedPricePerShare = yield.mulDivDown(1e18, yieldInShares);
```

Comment:
Why not use yield.divWadDown(yieldInShares)?

Reply:
yes that will work as well, don't mind the details, the point was to capture the intended behavior